### PR TITLE
debug-mode improvements + CommonServerPowerShell fix

### PR DESCRIPTION
**Pull request by glicht**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
* Better support for python logging being called before first debug-mode was called
* Better support for sensitive key replacement
* CommonServerPowerShell: removed Write-Output override and updated Write-Host override to use an alias.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Demisto
- [ ] 4.5.0
- [x] 5.0.0
- [ ] 5.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by glicht** [comment]
TDD. Test failing before fix: https://app.circleci.com/pipelines/github/demisto/content/13858/workflows/71b25737-aa01-4f14-bfc6-86c993c080b7/jobs/55219

image

File: .circleci/config.yml [hunk 1]
```diff
@@ -122,9 +122,9 @@ jobs:
                 exit 0
             fi
 
-            # Run flake8 on all excluding Integraions and Scripts (they will be handled in linting)
+            # Run flake8 on all excluding Packs (Integraions and Scripts) - they will be handled in linting
             ./Tests/scripts/pyflake.sh *.py
-            find . -maxdepth 1 -type d -not \( -path . -o -path ./Integrations -o -path ./Scripts -o -path ./Beta_Integrations -o -path ./venv \) | xargs ./Tests/scripts/pyflake.sh
+            find . -maxdepth 1 -type d -not \( -path . -o -path ./Packs -o -path ./venv \) | xargs ./Tests/scripts/pyflake.sh
```
Comment (Itay4) [inline .circleci/config.yml:127]: we should still run on `Beta_Integrations`

File: Packs/Base/Scripts/CommonServerPowerShell/CommonServerPowerShell.ps1 [hunk 2]
```diff
@@ -280,8 +280,12 @@ class DemistoObject {
...
 
 [DemistoObject]$demisto = [DemistoObject]::New()
-function global:Write-Host($UserInput) { $demisto.Log($UserInput) | Out-Null }
-function global:Write-Output($UserInput) { $demisto.Log($UserInput) | Out-Null }
+
+function global:Write-HostToLog($UserInput) { $demisto.Log($UserInput) | Out-Null }
+# we override Write-Host with a function which writes to the demisto.Log.
+# Incase there is need to undo this Alias, the integration/script can run the following command
+# Remove-Item 'Alias:Write-Host' -Force
+Set-Alias -Name 'Write-Host' -Value 'Write-HostToLog' -Scope Global
```
Comment (Itay4) [inline Packs/Base/Scripts/CommonServerPowerShell/CommonServerPowerShell.ps1:288]: why dont we need the same thing for `Write-Output` ?

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 3]
```diff
@@ -3214,26 +3220,44 @@ def __init__(self):
...
+            self.org_handlers.extend(self.root_logger.handlers)
+            for h in self.org_handlers:
+                self.root_logger.removeHandler(h)            
         self.root_logger.addHandler(self.handler)
 
     def __del__(self):
         if self.handler:
             self.root_logger.setLevel(self.prev_log_level)
-            self.root_logger.removeHandler(self.handler)
+            self.root_logger.removeHandler(self.handler)
```
Comment (Itay4) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:3233]: revert this

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 4]
```diff
@@ -3214,26 +3220,44 @@ def __init__(self):
...
             else:
                 delattr(self.http_client, 'print')
+    
+    def log_start_debug(self):
+        """
+        Utility function to log start of debug mode logging
+        """
+        msg = "debug-mode started.\nhttp client print found: {}.\nEnv {}.".format(self.http_client_print is not None, os.environ)
+        if hasattr(demisto, 'params'):
+            msg = msg + "\nParams: {}.".format(demisto.params())
```
Comment (Itay4) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:3252]:
```suggestion
            msg += "\nParams: {}.".format(demisto.params())
```

Comment (Itay4) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:3252]: this log will go through `_iter_sesistive_dict_obj` and secrets will be sanitized ?

File: Packs/Base/TestPlaybooks/script-TestPyLogging.yml [hunk 5]
```diff
@@ -0,0 +1,20 @@
+commonfields:
+  id: TestPyLogging
+  version: -1
+name: TestPyLogging
+script: |
+  import logging
+
+  root_logger = logging.getLogger()
+
+  logging.debug("this is a logging test")
```
Comment (Itay4) [inline Packs/Base/TestPlaybooks/script-TestPyLogging.yml:10]: any value to print to different values here? `logging.info` for example

File: .circleci/config.yml (hunk 1 above)
Comment (glicht) [reply .circleci/config.yml:127]: this line specified on what directories NOT to run. As we don't have this dir any more it is not needed.

File: Packs/Base/TestPlaybooks/script-TestPyLogging.yml (hunk 5 above)
Comment (glicht) [reply Packs/Base/TestPlaybooks/script-TestPyLogging.yml:10]: don't see a need for it

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py (hunk 4 above)
Comment (glicht) [reply Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:3252]: Yes. I just pushed also a unit test to test this out.

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 6]
```diff
@@ -966,15 +966,17 @@ def __init__(self):
...
-                    for p in sensitive_params:
-                        if p in k_lower:
-                            if isinstance(v, STRING_OBJ_TYPES):
-                                self.add_replace_strs(v, b64_encode(v))
-                            if isinstance(v, dict) and v.get('password'):  # credentials object case
-                                pswrd = v.get('password')
-                                self.add_replace_strs(pswrd, b64_encode(pswrd))
+                self._iter_sesistive_dict_obj(demisto.params(), sensitive_params)
+
+    def _iter_sesistive_dict_obj(self, dict_obj, sensitive_params):
```
Comment (anara123) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:971]:
```suggestion
    def _iter_sensitive_dict_obj(self, dict_obj, sensitive_params):
```

Comment (glicht) [reply Packs/Base/Scripts/CommonServerPython/CommonServerPython.py]: fixed

**Comment by kirbles19** [comment]
@glicht Done

**Comment by glicht** [comment]
force merge. failing unit test is not related.

**Pull request merged**